File: app/src/api/settings/services/service-service.ts
import { Service } from '@/api/settings/data-transfer-objects/service';
import { http } from '@/api/core/http';
import { ServiceCreate } from '@/api/settings/data-transfer-objects/service-create';
import { ServiceUpdate } from '@/api/settings/data-transfer-objects/service-update';
import { ServiceConfiguration } from '@/api/settings/data-transfer-objects/service-configuration';

// Don't think about the name to hard.
export class ServiceService {
    async getServices(): Promise<Service[]> {
        try {
            const res = await http.get('settings/service');

            if (res.data == null) {
                return [];
            }

            return res.data.map((s: any) => this._map(s));
        } catch (e) {
            if (e.response == null || e.response.status == 404) {
                return [];
            } else {
                throw e;
            }
        }
    }

    async createService(createService: ServiceCreate): Promise<Service> {
        const res = await http.post('settings/service', createService);
        const s = this._map(res.data);

        return s;
    }

    async updateService(updateService: ServiceUpdate): Promise<Service> {
        const res = await http.put(`settings/service/${updateService.id}`, updateService);
        const s = this._map(res.data);
        return s;
    }

    async deleteService(id: string) {
        await http.delete(`settings/service/${id}`);
    }

    _map(d: any): Service {
        const s = new Service(
            d.id,
            d.name,
            d.description,
            d.pricingMethod,
            d.configurations.map((c: any) => {
                const config = new ServiceConfiguration(c.id, c.vehicleCategoryId, c.price, c.duration);
                config.id = c.id;

                return config;
            })
        );
        return s;
    }
}

File: api/src/Domain/Scheduling/Settings/Services/ServiceService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using DetailingArsenal.Domain.Users;

namespace DetailingArsenal.Domain.Settings {
    public interface IServiceService : IService {
        Task<Service> GetById(Guid id);
        Task<List<Service>> GetByUser(User user);
        Task<Service> Create(ServiceCreate create, User user);
        Task Update(Service service, ServiceUpdate update);
        Task Delete(Service service);
    }

    public class ServiceService : IServiceService {
        IServiceRepo repo;
        ServiceNameUniqueSpecification uniqueNameSpec;
        ServiceNotInUseSpecification notInUseSpec;

        public ServiceService(IServiceRepo repo, ServiceNameUniqueSpecification uniqueNameSpec, ServiceNotInUseSpecification notInUseSpec) {
            this.repo = repo;
            this.uniqueNameSpec = uniqueNameSpec;
            this.notInUseSpec = notInUseSpec;
        }

        public async Task<Service> GetById(Guid id) {
            return await repo.FindById(id) ?? throw new EntityNotFoundException();
        }

        public async Task<List<Service>> GetByUser(User user) {
            return await repo.FindByUser(user);
        }

        public async Task<Service> Create(ServiceCreate create, User user) {
            var service = Service.Create(
                user!.Id,
                create.Name,
                create.Description,
                create.PricingMethod
            );

            service.Configurations = create.Configurations.Select(
                c => ServiceConfiguration.Create(
                    service.Id,
                    c.VehicleCategoryId,
                    c.Price,
                    c.Duration
                )
            ).ToList();

            await uniqueNameSpec.CheckAndThrow(service);
            await repo.Add(service);

            return service;

        }
        public async Task Update(Service service, ServiceUpdate update) {
            service.Name = update.Name;
            service.Description = update.Description;
            service.PricingMethod = update.PricingMethod;

            service.Configurations = update.Configurations.Select(
                c => ServiceConfiguration.Create(
                    service.Id,
                    c.VehicleCategoryId,
                    c.Price,
                    c.Duration
                )
            ).ToList();

            await uniqueNameSpec.CheckAndThrow(service);
            await repo.Update(service);
        }

        public async Task Delete(Service service) {
            await notInUseSpec.CheckAndThrow(service);
            await repo.Delete(service);
        }
    }
}